Trait method impl restrictions

joshtriplett opened the pull request:
Support restricting implementation of individual methods within traits, as an
extension of the restriction mechanism defined in
[RFC 3323](https://rust-lang.github.io/rfcs/3323-restrictions.html).

This makes it possible, for instance, to define a trait that any crate can
implement, but disallow overriding one of the trait's methods.

This was inspired in the course of writing another RFC defining a trait, which needed precisely this feature of restricting overrides of the trait's method. I separated out this feature as its own RFC, since it's independently useful for various other purposes, and since it should be available to any crate and not just the standard library.

burdges commented:
We've past proposals for inherent trait methods I think, mostly proposing a second inherent looking `impl` block, not sure that syntax matters, but this winds up functionally equivelent, yes?

There is however a question of default speed vs size optimizations in vtables, aka do these methods appear in the vtable, or do they have a generic implementation for every type?  Also, how does one flag that these method should appear in a vtable, or should use a generic implementation for every type?

Lokathor commented:
It might be better to have this as an attribute rather than yet another potential keyword position for parsing to deal with. Otherwise, great.

joshtriplett commented:
@Lokathor wrote:
> It might be better to have this as an attribute rather than yet another potential keyword position for parsing to deal with. Otherwise, great.

I get that, but on the flip side, that'd be inconsistent with RFC 3323, and it seems awkward to use a keyword in one place and an attribute in another.

joshtriplett commented:
@burdges wrote:
> We've past proposals for inherent trait methods I think, mostly proposing a second inherent looking `impl` block, not sure that syntax matters, but this winds up functionally equivelent, yes?

That might be equivalent to a subset of this, depending on the details. I haven't seen those proposals.

> There is however a question of default speed vs size optimizations in vtables, aka do these methods appear in the vtable, or do they have a generic implementation for every type? Also, how does one flag that these method should appear in a vtable, or should use a generic implementation for every type?

In theory, if you have a method that can't be overridden outside of the crate/module, and nothing overrides it, you could optimize by omitting it from the vtable. I don't think that optimization should be mandatory, or required for initial implementation, though.

burdges commented:
I'm more asking what's the best default.  If the best default were to be in the vtable, then you can just do
```
trait Trait {
    #[inline(always)]
    impl(crate) fn f(&self) { f_inner(self) }
}
fn f_inner<T: Trait>(s: &T) { .. }
```
I'd expect `f_inner` gets only one copy for the trait obejct here.

If otoh the best default were not to be in the vtable, then rustc should do something more complex.

Anyways yeah maybe not relevant here.

bluebear94 commented:
Another future possibility could be to add `impl(unsafe)` trait methods, which can only be (re)implemented by unsafe impls.

joshtriplett commented:
@bluebear94 Interesting! So, rather than requiring `unsafe impl Trait for Type`, you could implement the trait safely, as long as you don't override the method? That's a novel idea.

programmerjake commented:
this would be quite useful for stabilizing `std::error::Error::type_id`, which currently has several workarounds to prevent it from being overridden:
https://doc.rust-lang.org/1.80.1/src/core/error.rs.html#88-100
cc rust-lang/rust#60784

joshtriplett commented:
@programmerjake That's a great use case, thank you!

joshtriplett commented on text/3678-trait-method-impl-restrictions.md:
````diff
@@ -0,0 +1,145 @@
+}
+```
+
+Note that in some cases, the compiler may be able to avoid placing an
+`impl`-restricted method in the trait's vtable, if there is only a single
+implementation that can ever be invoked.
````
@scottmcm I don't think I understand how that case relates. We can't do this for `RangeBounds` because there are multiple implementations of `RangeBounds`.

If we have a trait where there can only be one implementation of a given method, that effectively lets us do a trivial version of the devirtualization optimization: call the one and only implementation directly, no indirection.

joshtriplett replied:
@scottmcm I see; thanks for the explanation. Based on that example, I agree that there are cases where the compiler may prefer to monomorphize a method and put it into the vtable.

I think it's still useful for the RFC to observe that the compiler *may* wish to devirtualize, if the compiler decides that the one-and-only implementation *doesn't* benefit from monomorphization. But I'll reword it to clarify and avoid the implication that doing so would always be the right thing if possible.

burdges commented:
I think `#[final]` is easier to explain too.  As usages of the `impl` and `use` keywords multiply, we quickly lose our ability to explain them.

tgross35 commented:
If it is to be an attribute then I would prefer something like `#[no_override]` that says what it does, rather than taking a C++ keyword that imo isn't very descriptive. But +1 to this suggestion - I think it is immediately obvious what the effects are, and either being on or off is easier to follow than giving fine grained control. Agreeing with Scott, being able to e.g. override something in the module but not the rest of the crate doesn't seem like a common enough need to justify the complexity.

I assume the syntax was chosen for parity with [RFC3323](https://rust-lang.github.io/rfcs/3323-restrictions.html), but I think it is okay to deviate from this if it comes with a simplification.

joshtriplett commented:
@scottmcm
> If we're not adding pub() fn foo() {} and struct Foo { mut() a: i32 }, then I think any superficial similarity here is more harmful than good.

I think we *should* add `mut()` or equivalent: "this field is not writable from anywhere". (Useful for an object ID or similar that can never change once constructed.)

I agree that `pub()` ("this can't be called from anywhere") isn't useful.

> If we need a special syntax, let's make it final or #[final].

While I do personally think the RFC 3678 syntax has value, I'm not going to continue proposing that that syntax when we don't have a concrete use case for the cases other than `impl()`. I've changed the RFC to propose `final`, and moved the generalized RFC-3678-style syntax to future possibilities.

joshtriplett commented:
Canceling and restarting the FCP so that anyone who previously checked their box can decide if they still wish to do so with the new version that uses `final`.

@rfcbot cancel

@rfcbot merge

tmandry commented:
Makes sense to me. Sometimes provided methods are awkward because the thing you really want is a final method (thinking about the `super_*` visitor methods in the compiler, which are called from overrideable methods). Thanks for proposing this.

@rfcbot reviewed

kennytm commented on text/3678-final.md:
```diff
@@ -0,0 +1,123 @@
+function outside a crate while allowing implementations within the crate.
+(Likewise with `impl(self)` or any other visibility.)
+
+We could theoretically allow `final` restrictions on associated types, as well.
```
and associated const as well? 

```suggestion
We could theoretically allow `final` restrictions on associated consts and types, as well.
```

something like

```rust
pub trait GenericArray<T, const N: usize> {
    final type Array = [T; N];
    final const SIZE: usize = std::mem::size_of::<T>() * N;
}
```

nikomatsakis commented:
I have wanted this feature many, many times over the years. It would also allow the methods defined (rather awkwardly) as inherent methods on dyn Any to be moved to the trait, allowing them to be called on subtraits of any. 

@rfcbot reviewed

nikomatsakis commented:
That said,  a design meeting plus some time for folks to comment both seem good.

One possible extension: final associated types, which I have also wanted. I have to go digging around to see why exactly:)

nikomatsakis commented:
One example:

It would be nice if IntoIterator were defined as

```
trait IntoIterator {
    final type Item = Self::Iter::Item;
    type Iter: Iterator;
    fn into_iter(self) -> Self::Iter;
}
```

bluebear94 commented:
Associated constants could also be made final as a further possibility.

programmerjake commented:
> Associated constants could also be made final as a further possibility.

imo `final` should be supported on any item you can put in a trait definition:
associated types
associated constants
functions
anything else we add in the future.

tgross35 commented on text/3678-final.md:
```diff
@@ -0,0 +1,123 @@
+We can always add such syntax for the more general cases in the future if
+needed; see the future possibilities section.
+
+We could use `#[final]` rather than `final`. However, since we already have the
+`final` keyword reserved, using that keyword seems syntactically simpler than
+an attribute.
```
The motivation here doesn't seem strong enough here to justify new syntax. As far as how this affects API consumers, it seems more like a logical continuation of `#[must_use]`, `#[non_exhaustive]`, `#[deprecated]`, or `#[diagnostic::*]` - and it doesn't seem to be more likely to be used.

Also - is `final` intuitive enough for anyone not coming with C++ or Java? I found the keyword somewhat unintuitive in C++ and I don't think consistency with those languages is worth enough to not do something like `#[no_override]`, which is much more obvious (similar to `non_exhaustive`).

knickish replied:
The `final` keyword is also used in Swift

joshtriplett replied:
I've reworded this to avoid coming across as a rationale, rather than just a description of an alternative.

tmandry commented on text/3678-final.md:
```diff
@@ -0,0 +1,123 @@
+
+The keyword `final` has been reserved since Rust 1.0, so this feature can ship
+identically in all editions.
+
```
It should also state that `final` has no impact on dyn compatibility of the trait.

At first I thought it would, but I realized that it shouldn't.. if you have a self-by-value method that's not going to work with `dyn Trait` as the self type regardless of whether the method is final. So you still need `where Self: Sized` (and in this particular case, a bridging `impl Trait for Box<dyn Trait>` automatically gets you back the method).

joshtriplett replied:
@scottmcm I tend to remove that section (and retitle the remaining one as "Explanation") when there isn't any meaningful difference in content for it. But the items that you're suggesting seem like reasonable additions to me. Thanks! I'll merge and then do some edits.

jhpratt commented:
> I get that, but on the flip side, that'd be inconsistent with RFC 3323, and it seems awkward to use a keyword in one place and an attribute in another.

Counterpoint: the syntax in the restrictions RFC is an unresolved question. My personal preference has grown towards a `#[restrict(impl(crate))]` sort of syntax, which solves other issues re. pre-existing macros as well.

joshtriplett commented:
@jhpratt
> > I get that, but on the flip side, that'd be inconsistent with RFC 3323, and it seems awkward to use a keyword in one place and an attribute in another.
> 
> Counterpoint: the syntax in the restrictions RFC is an unresolved question. My personal preference has grown towards a `#[restrict(impl(crate))]` sort of syntax, which solves other issues re. pre-existing macros as well.

Since my original comment here, I've switched to no longer attempting to mirror the RFC 3323 syntax.

joshtriplett commented:
@scottmcm I believe I've now addressed your concern.

nikomatsakis commented:
Something I realized:

@joshtriplett it's worth mentioning, I think, that we could allow marker traits to have final methods (but not other methods).

PoignardAzur commented:
Maybe a little late, but I'd like to suggest another syntax:

```rust
trait MyTrait: Display {
    // stuff
}

impl MyTrait {
    fn method(&self) {
        println!("MyTrait::method: {self}");
    }
}
```

This syntax would basically be an inherent impl block, for traits. It would have the same properties as `final` in this RFC.

kpreid commented:
Interesting. `impl MyTrait {}` used to be valid syntax, with the meaning that is today given to `impl dyn MyTrait {}`, before `dyn` was mandatory for trait object types. This argues both against and in favor of it:

* Against: It might be confusing that an old syntax has been given a new but different meaning.
* For: The new meaning is one that I recall a lot of beginners *wanting*, and they had to be told that their program didn't do what they hoped (that the methods would be found only if a `dyn` type was involved and not otherwise).

Other arguments:

* For: This makes it very straightforward that such a method is not something a trait impl can override because it's *not in the `trait` declaration* — it uses an existing principle about what goes in an `impl`, rather than adding a new kind of access restriction.
* For: No new grammar whatsoever (`impl MyType {}` and `impl MyTrait {}` are the same at the syntactic level).
* For: This adds all the code-organization power of `impl` blocks:
    * Putting the blocks anywhere you want instead of embedding all the final methods in the `trait` declaration itself
    * Declaring bounds on the block instead of the individual method or the trait
* Against: Now it's harder to see what methods importing a trait introduces (when reading source code, not documentation).

Personally, I don't care very much — I'm looking forward to being able to use `final`ness regardless of what syntax it has. Both seem elegant enough.

PoignardAzur commented:
You could add the keyword to those impl blocks.

slanterns commented:
I have conflict feelings towards the newly proposed syntax, i.e. "just think it like inherent methods on types!" vs "all of the trait's contract should be visible from the trait definition itself!" :(

(For the latter, a final method, though cannot be overrided, is also a part of the trait's "contract" just like all other ordinary methods.)

kpreid commented:
Regarding contracts, here is a hypothetical future feature that would build on `final`:

1.  In general, a trait serves two users: callers and implementors.
2.  Traits have properties which all uses of the traits should obey; these properties are currently only set down in signatures (checked) where possible, and documentation (not checked) otherwise.
3.  `final fn`s allow traits to provide methods which are guaranteed to do validation in between the caller and the implementor, checking properties or enforcing them via transformation of the data/calls. Thus, the caller can rely on certain properties without relying on the implementor.
4.  However, there's a further missing piece: a caller can always call any trait function without such validation. We could imagine another feature which is functions which are

    * implementable (not `final`),
    * callable by code in the module defining the trait (or some other visibility scope; details could vary), and
    * *not* callable by other code (that is merely using the trait).

    Such functions would provide a guarantee that an *implementor* can rely on, that they are only called in accordance with the trait’s rules.

In this lens, the trait can end up having *two different* API surfaces: one for callers and one for implementors. When I started writing, I thought this was a good argument for `final fn`, but now I think it isn't, really. But maybe it will inspire some more useful thoughts.

bluebear94 commented:
> * _not_ callable by other code (that is merely using the trait).

If you wanted to restrict calling a trait method to the containing module, then you could make it take an instance of a type that can only be constructed in the module:

```rust
pub struct AToken(());

pub trait A {
    fn called_privately(&self, a: u32, b: &str, token: AToken) -> bool;
}

// somewhere later in self:
fn do_something(a: &impl A) {
    let _ = a.called_privately(5, "test", AToken(()));
}
```

kpreid commented:
That constrains the _maximum count_ of calls to `called_privately()` only; an implementor can then call another implementation’s `called_privately()` with arbitrary `a` and `b`. It also can’t enforce any rules about the pattern of calls to multiple methods unless the token gets involved in more complex ways. But let’s not go too far along the tangent of designing or debating the value of that feature; I only mentioned it to see if it provided any inspiration for the choice of syntax for `final`-or-whatever.

cynecx commented:
@traviscross I was slightly confused for a moment when I read:

> Write this to prevent the "override" below?

I think using the "overriding" terminology here is a bit ambiguous because you are technically not overriding anything. It's perfectly valid to have a same named trait method in `Cat` and `Mammal`, and that's probably not going to change? A trait bound like in `trait Cat where Self: Mammal` surely shouldn't introduce any restrictions in what can be defined in the `Cat` trait.

I think a better way to describe that would be to simply say "to prevent shadowing" because that's what #3624 is doing (allowing shadowing).

ijackson commented:
I think the word "inherent" is defined in the Reference to mean methods that are defined in an types's own `impl`.  https://doc.rust-lang.org/reference/items/implementations.html#inherent-implementations

The result is that the word "inherent" has come to mean "a method defined on the type itself, not on some trait that it implements".

Therefore this is the wrong term to use for this feature.

joshtriplett commented:
This has an analogous meaning: "method defined by the trait itself, not by any instance of the trait".

ijackson commented:
> This has an analogous meaning: "method defined by the trait itself, not by any instance of the trait".

Yes, I can see the analogy.  However, this is very confusing, and a different term would be much better.  This breaks a whole lot of existing documentation (both within the rust-lang org, and elsewhere, and presumably error messages) that talks about "inherent methods".

ijackson commented:
To expand on that: it's all very well to have some analogy that justifies the usage, but what really matters is qutestions like:

 1. What the central meaning of the word "inherent" now?  Answer: *specifically not* a trait method
 2. How will this affect how we teach Rust?  Answer: we will have to change a lot of existing teaching materials, and it will generate a lot of confusion.
 3. What term should be used for the current meaning of "inherent" (ie "not a trait method")?  Answer: this RFC does not answer this question, and should not be accepted for that reason alone.
 4. Is there a better term we could use for this new feature?  Yes, "final" was already suggested, and there are other possibilities

runiq commented:
Coming here from TWIR, I also like `#[final]` better.FWIW In my eyes, both types and traits have equal claim to the meaning of 'inherit', as @ijackson has laid down above. Also, while 'final' does a related-but-not-exactly-equivalent feature in C++, couldn't that be useful as a jumping-off point in the documentation? As in, the docs would have to focus mostly on the differences?

tmandry commented:
I thought we were going to restart proposed FCP. I'm not that happy with `#[inherent]` and at this point wouldn't check my box on it, so I think we should restart.

@rfcbot concern proposed FCP should be restarted with a new name

Looking over the comments, I think there are two "attractor states" I'm comfortable with:

1. Using `final` or `#[final]` – This is fine because it means something very analogous to its meaning in other languages. As @scottmcm said, the _intent_ of the user is the same in both cases.
2. "Anonymous extension traits" – This needs more design work. I would like it to fit into an overall picture where we allow inherent implementations of traits. It isn't clear to me how it would extend to extension traits defined from outside the defining module of the original trait (how would you import the extension?). If we use `impl Trait { ... }` for this it also has the disadvantage of not having an obvious name to use when talking about it, as @scottmcm mentioned.

Given the open questions on the second I would just go with the first for now.

Extending the meaning of "inherent" from "inherent to a type" to "inherent to a trait" is something one can logically do, but it isn't clear to me what we gain from that. "Inherent" is already a term not that many users know, since there's no syntax for it, and the existing intuition they do have will have to be modified to make this term work. In the meantime there will be user confusion and we will have to spend a lot of time explaining what it means. That doesn't seem worth it to me when we have a word that users will already intuitively understand.

> That is, there already is this way of "extending" traits with subtraits, and there will soon be this other axis for the "overriding" of a method. This other axis is a better fit for final. Which isn't to say whether or not we'll ever want the knob on that axis, just that it makes it a poor fit here.

I really don't agree. Shadowing a method name is not analogous to overriding a method. They are completely different methods, and shadowing does not prevent access to the original method.

The intent of the feature in #3624 is to allow _supertraits_ to add method names that are already defined by subtraits. If the goal is to prevent this mechanism from being abused I wouldn't want a world where best practice is to declare every method in a new trait as `#[final]` – I'd rather flip the default and say that newly added methods ought to be declared `#[shadowable]`, or similar.

kennytm commented:
:face_exhaling:  can the keyword bikeshedding be deferred until before stabilization?

tmandry commented:
Procedurally I don't think we should make a major change after the other team members have already checked their box, unless they've affirmed that they are comfortable with the change.

Procedurally the bikeshed could be deferred, as in the keyword could be moved to an open question in the RFC, but practically I don't see what we would gain from that. The feature has straightforward semantics and the debate is not likely to be resolved through experience using it. The choice of how to expose this feature depends on differing mental models and a coherent vision for the language. That's the stuff of RFCs.

joshtriplett commented:
@tmandry :+1: for restarting the FCP, as you've done.

And FWIW, I personally agree and prefer `#[final]`, and would be happy to change the RFC back to that. I have no particular attachment to either name, and I'm happy to go with whichever has consensus amongst the team.

joshtriplett commented:
@traviscross I'm supportive of the concept of `impl Trait { ... }` blocks, but I'm still inclined to spell *this* feature `#[final]`, for simplicity. The more I think about it, the more I think it's not worth spending weirdness budget on adding another meaning for "inherent".

bstrie commented:
As someone whose default policy is "read my lips: no new syntaxes" I'm on board with the attribute. We can always make it a keyword in the future if there's sufficient demand for it.

kennytm commented:
`final` is already a reserved keyword.

```
error: expected identifier, found reserved keyword `final`
 --> src/main.rs:2:9
  |
2 |     let final = 4;
  |         ^^^^^ expected identifier, found reserved keyword
  |
help: escape `final` to use it as an identifier
  |
2 |     let r#final = 4;
  |         ++
```

joshtriplett commented on text/3678-final.md:
```diff
@@ -0,0 +1,179 @@
+method in the trait's vtable, if the one-and-only implementation does not
+benefit from monomorphization.
+
+Note that removing a `final` restriction is always forwards-compatible.
```
I can clarify that. But that's not new; removing a default implementation is already not backwards-compatible.

scottmcm commented on text/3678-final.md:
```diff
@@ -0,0 +1,184 @@
+One example of a trait in the standard library benefiting from this:
+`Error::type_id`, which has thus far remained unstable because it's unsafe to
+override. This RFC would allow stabilizing that method so users can call it,
+without permitting reimplementation of it.
```
Another thing that came to mind:
`````suggestion
One example of a trait in the standard library benefiting from this:
`Error::type_id`, which has thus far remained unstable because it's unsafe to
override. This RFC would allow stabilizing that method so users can call it,
without permitting reimplementation of it.

Another is that it might be nice to have
```rust
trait Copy : Clone {
    final fn copy(&self) -> Self { *self }
}
```
where the `final` is important in letting `unsafe` code trust
that it does exactly that, never anything else.

(Today if it was added, code would have to think about what would happen
if someone overrode it to do something else -- maybe with side effects, maybe 
to panic, etc.  By being `final` that's not a concern.)
`````

Feel free to take or adapt or ignore as you see fit.

scottmcm commented on text/3678-final.md:
```diff
@@ -0,0 +1,184 @@
+vtable. A design for this should take into account cases where people *do* want
+`final` methods to appear in the vtable, because they benefit substantially
+from monomorphization.
+
```
```suggestion

We could allow `final fn` methods on `#[marker]` traits, which are currently not allowed to have any methods (because they can't allow different implementations in different `impl`s).

```

Nadrieril commented on text/3678-final.md:
```diff
@@ -0,0 +1,184 @@
+As with any language feature, this adds more surface area to the language.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```
Rather than a "Future Possibility", I see my ["Explicit API Knobs"](https://nadrieril.github.io/blog/2025/11/12/flexible-explicit-api-knobs.html) proposal as an alternative.

Admittedly I don't mind `final` too much because it's unambiguous, but overall in the language I would like to avoid not-very-extensible keyword-based syntaxes like RFC 3323 ("Restrictions") and replace them with a common attribute language. If you wanted to include a paragraph on this, I'd propose:

> Rather than using a keyword, we could use an attribute as part of a common language for API control (https://nadrieril.github.io/blog/2025/11/12/flexible-explicit-api-knobs.html). This cool look like (strawman syntax):
```rust
trait Trait {
  #[forbid(downstream_crates(override))]
  fn method() { ... }
}
```
> The benefit is expressivity: there are other things we may want to control about a method, such as calling it (`#[forbid(downstream_crates(call))]`) or turbofishing it (`#[allow(future_api(add_generic_params))]`). Having an extensible language leaves room for future knobs like this.

jhpratt replied:
fwiw the syntax for restrictions was/is an unresolved question. I wouldn't mind moving towards `#[restrict(impl(crate))]` for example, being more general towards things like this.

Nadrieril replied:
I don't feel represented by the current text, FYI. The current text does mention RFC 3323 but only as an alternative keyword syntax. The aspect I care about is that API control be based on reusable+composable concepts. Neither `final` nor `impl(visibility)` reach that bar in my opinion.

Nadrieril replied:
That is quite funny x)

zachs18 commented on text/3678-final.md:
```diff
@@ -0,0 +1,200 @@
+`final` is only allowed in trait definitions. `final` is not allowed on impls
+or their items, non-trait functions, or `extern` blocks.
+
+A `final fn` has no impact on the `dyn`-compatibility of a trait.
```
The RFC also mentions `std::error::Error::type_id` as a possible use-case, but if `final fn` are not included in the vtable, then IIUC making `Error::type_id` `final` using the [current implementation](https://github.com/rust-lang/rust/blob/main/library/core/src/error.rs#L126) would be incorrect (it returns `TypeId::of::<Self>()`, not `self.type_id()`, so would give the wrong result for `dyn Error + 'static` if it doesn't go through the vtable).

At the least, it might make sense to explicitly call this out in the documentation of `final fn`, whichever way this ends up (`final fn` never going through vtable or affecting dyn-compatibilty, when called on `dyn Trait`, the method body "sees" `Self = dyn Trait`; or `final fn` behaving the same as `fn` w/r/t dyn-compatbility, when called on `dyn Trait`, the method body "sees" `Self = TheConcreteErasedType`; or something else). 

(The "Future Possibilities" portion of the RFC also says "We could add the optimization of not putting final methods in the trait vtable", but as the example of `Error::type_id` shows, whether a `final fn` is monomorphized with `Self = TheConcreteErasedType` or `Self = dyn Trait` matters for the semantics of the method body, so I think it is not just an "optimization" in all cases)

tmandry commented:
Looks good. Thanks for pushing this forward @joshtriplett.

@rfcbot reviewed
